#86639 but based on #87067

postamar opened the pull request:
This is #86639, but based on #87067.

Release justification: not for release
Release note: None

ajwerner committed:
sql/schemachanger: add back TXN_DROPPED, stop using OFFLINE

We added the `OFFLINE` and performed write operations during drops to move the
descriptor into that state in order to deal with the fact that synthetically
dropped descriptors would not appear so in virtual tables. Fortunately, this
shortcoming has been rectified. We now revive the TXN_DROPPED state. This
proves to be important because the `OFFLINE` state is not backwards compatible,
and the `TXN_DROPPED` state is. Ideally we'd find a way to move the descriptors
straight to `DROPPED` in the pre-commit case for all the cases we today
support, but doing that would require an overhaul of the dependency rules
which,at this point in the cycle, feels risky.

Fixes #86626

Release justification: Fixes a release-blocking backwards incompatibility.

Release note: None

ajwerner committed:
sql/schemachanger/scplan/internal/scstage: accept non-revertible in PreCommit

We can schedule non-revertible operations into PreCommit so long as the next
stage is PostCommitNonRevertible.

Release justification: part of bug fix

Release note: None

ajwerner committed:
sql/schemachanger/scplan/internal/scstage: implicitly satisfy no-op edges

Release note: None

ajwerner committed:
sql/schemachanger/rel: internally store rules with random vars

This is just to avoid bugs later.

Release justification: cleanup

Release note: None

ajwerner committed:
sql/schemachanger/rel: support not-join subqueries

This patch endows rel with the ability to define not-join rules which allow
extra variables to be introduced inside the clauses. When a not-join rule is
invoked, it leads to the outer query finding a contradiction if the rule fully
unifies.

Fixes #77148

Release justification: part of a broader bug-fix

Release note: None

ajwerner committed:
sql/schemachanger: enforce online schema change invariants with dep rules

Previously we had implicit rules in opgen which were used to enforce
sequencing between transitions. This was brittle and made it hard to
collapse certain edges. Now that we can represent a not-join, we can
add the rules we want selectively, only when descriptors are not being
dropped.

Fixes #86691
Fixes #86626

Release justification: fixes a bug

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87223)
<!-- Reviewable:end -->

Commit:
fixup! backup test expected results

Probably belongs to: add back TXN_DROPPED, stop using OFFLINE

Release justification: None
Release note: None

Commit:
fixup! TestExecBuild expected results

This possibly belongs in the last commit, I'm not sure.

Release justification: None
Release note: None

Commit:
fixup! bad fmt string in rel error msg

This belongs in: support not-join subqueries

Release justification: None
Release note: None

Commit:
fixup! synthetic descriptor handling

The code changes belong in: add back TXN_DROPPED, stop using OFFLINE
The test output changes I'm not so sure about.

Release justification: None
Release note: None

postamar commented:
@ajwerner I took your branch and rebased it on master after https://github.com/cockroachdb/cockroach/issues/87067 merged. My changes are captured in the "fixup!" commits. These mostly do not include changes that I requested when reviewing your PR, those remain to be done.

At this point would probably be simpler for you and perhaps for everyone to just squash the whole stack and instead talk about the distinct changes in the message of the squashed commit.

ajwerner commented:
This has been subsumed into #86639.
